src/modm/platform/i2s/stm32/i2s_master.hpp.in

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2021, Marton Ledneczki
+ * Copyright (c) 2021-2022, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_STM32_I2S_MASTER{{ id }}_HPP
+#define MODM_STM32_I2S_MASTER{{ id }}_HPP
+
+#include <cstdlib>
+
+#include <modm/architecture/interface/i2s_master.hpp>
+#include <modm/architecture/interface/peripheral.hpp>
+#include <modm/platform/gpio/connector.hpp>
+#include <modm/math/algorithm/prescaler.hpp>
+#include "i2s_hal_{{ id }}.hpp"
+
+namespace modm::platform
+{
+
+/**
+ * Inter-IC Sound (I2C{{ id }}).
+ *
+ * @tparam	DmaChannelTX DMA channel for sending
+ *
+ * @author	Marton Ledneczki
+ * @author	Raphael Lehmann
+ * @ingroup	modm_platform_i2s modm_platform_i2s_{{id}}
+ */
+template <class DmaChannelTx>
+class I2sMaster{{ id }} : public modm::I2sMaster, public I2sBase
+{
+	struct Dma {
+		using TxChannel = typename DmaChannelTx::template RequestMapping<Peripheral::Spi{{ id }}, DmaBase::Signal::Tx>::Channel;
+
+		static constexpr DmaBase::Request TxRequest = DmaChannelTx::template RequestMapping<Peripheral::Spi{{ id }}, DmaBase::Signal::Tx>::Request;
+	};
+
+private:
+	using Hal = I2sHal{{ id }};
+
+public:
+	template< class... Signals >
+	static void
+	connect()
+	{
+		using Connector = GpioConnector<Peripheral::I2s{{ id }}, Signals...>;
+		using Ck = typename Connector::template GetSignal<Gpio::Signal::Ck>;
+		using Mck = typename Connector::template GetSignal<Gpio::Signal::Mck>;
+		using Sd = typename Connector::template GetSignal<Gpio::Signal::Sd>;
+		using Ws = typename Connector::template GetSignal<Gpio::Signal::Ws>;
+
+		Ck::setOutput(Gpio::OutputType::PushPull);
+		Mck::setOutput(Gpio::OutputType::PushPull);
+		Sd::setOutput(Gpio::OutputType::PushPull);
+		Ws::setOutput(Gpio::OutputType::PushPull);
+		Connector::connect();
+	}
+
+	template<class SystemClock, frequency_t samplerate, percent_t tolerance = pct(0.019)>
+	static inline void
+	initialize(BitDepth bitDepth = BitDepth::SixteenWithChannel16,
+			   MasterClockOutput masterClockOutput = MasterClockOutput::Enabled,
+			   I2sStandard i2sStandard = I2sStandard::Philips,
+			   ClockPolarity clockPolarity = ClockPolarity::Low,
+			   DmaBase::Priority dmaPriority = DmaBase::Priority::High)
+	{
+		constexpr float prescaler = static_cast<float>(SystemClock::I2sPll) / (samplerate*16*2*8);
+		constexpr uint8_t i2s_div = static_cast<uint8_t>(prescaler/2);
+		constexpr uint8_t odd_factor = ((prescaler - static_cast<float>(i2s_div*2)) >= 0.5) ? 1 : 0;
+		constexpr float real_samplerate = static_cast<float>(SystemClock::I2sPll) /
+											(16*2*8*(2*i2s_div+odd_factor));
```

Comment:
Oh yes. I added the bitDepth parameter, but did not thought about the prescaler computation.
I'll push a fix soon.

Reply:
It is actually even more complicated, because the master clock output also has an influence on the prescaler (the magic `*8` multiplication in the previous code):

![image](https://user-images.githubusercontent.com/2820734/159543197-783e1cf4-e656-477a-b662-83de77dcf98c.png)